File: backend-ruby/trino_client.rb
require "base64"
require "json"
require "net/http"
require "uri"

class TrinoClient
  def initialize(
    base_url,
    user: ENV.fetch("TRINO_USER", "log_search"),
    password: ENV.fetch("TRINO_PASSWORD", ""),
    catalog: ENV.fetch("TRINO_CATALOG", "iceberg"),
    schema: ENV.fetch("TRINO_SCHEMA", "logs")
  )
    base = base_url.end_with?("/") ? base_url : "#{base_url}/"
    @statement_uri = URI.join(base, "v1/statement")
    @user = user
    @password = password
    @catalog = catalog
    @schema = schema
  end

  def ping
    execute("SELECT 1", timeout: 5)
    true
  rescue StandardError
    false
  end

  def execute(sql, timeout: 15)
    response = request(:post, @statement_uri, timeout: timeout, body: sql)
    collect_pages(JSON.parse(response.body), timeout: timeout)
  end

  private

  def collect_pages(body, timeout:)
    rows = []
    columns = []

    loop do
      if body["error"]
        message = body["error"]["message"] || body["error"].to_s
        raise "Trino query failed: #{message}"
      end

      rows.concat(body.fetch("data", []))
      columns = body["columns"].map { |column| column["name"] } if body["columns"] && columns.empty?

      next_uri = body["nextUri"]
      return [rows, columns] unless next_uri

      response = request(:get, URI(next_uri), timeout: timeout)
      body = JSON.parse(response.body)
    end
  end

  def request(method, uri, timeout:, body: nil)
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = uri.scheme == "https"
    http.open_timeout = timeout
    http.read_timeout = timeout

    request = method == :post ? Net::HTTP::Post.new(uri) : Net::HTTP::Get.new(uri)
    trino_headers.each { |key, value| request[key] = value }
    request.body = body if body

    response = http.request(request)
    raise "Trino HTTP #{response.code}: #{response.body}" unless response.is_a?(Net::HTTPSuccess)

    response
  end

  def trino_headers
    headers = {
      "X-Trino-User" => @user,
      "X-Trino-Source" => "ruby-sinatra-trino-log-search",
      "Content-Type" => "text/plain; charset=utf-8"
    }
    headers["X-Trino-Catalog"] = @catalog unless @catalog.empty?
    headers["X-Trino-Schema"] = @schema unless @schema.empty?
    headers["Authorization"] = "Basic #{Base64.strict_encode64("#{@user}:#{@password}")}" unless @password.empty?
    headers
  end
end

File: backend-java/src/main/java/com/example/flasktrino/TrinoClient.java
package com.example.flasktrino;

import com.fasterxml.jackson.core.type.TypeReference;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;
import java.util.Map;

final class TrinoClient implements App.QueryClient {
    private final App.Config config;
    private final HttpClient client = HttpClient.newBuilder()
            .connectTimeout(java.time.Duration.ofSeconds(10))
            .build();
    private final URI statementUri;

    TrinoClient(App.Config config) {
        this.config = config;
        this.statementUri = URI.create(config.trinoUrl().replaceAll("/+$", "") + "/v1/statement");
    }

    @Override
    public boolean ping() {
        try {
            execute("SELECT 1");
            return true;
        } catch (Exception ignored) {
            return false;
        }
    }

    @Override
    public App.QueryResult execute(String sql) throws Exception {
        HttpRequest.Builder builder = HttpRequest.newBuilder(statementUri)
                .timeout(java.time.Duration.ofSeconds(60))
                .POST(HttpRequest.BodyPublishers.ofString(sql, StandardCharsets.UTF_8));
        applyHeaders(builder);
        HttpResponse<String> response = client.send(
                builder.build(),
                HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8)
        );
        if (response.statusCode() < 200 || response.statusCode() >= 300) {
            throw new IOException("trino statement failed: HTTP " + response.statusCode() + ": " + response.body());
        }
        return collect(App.JSON.readValue(response.body(), new TypeReference<>() {
        }));
    }

    private App.QueryResult collect(Map<String, Object> page) throws Exception {
        List<List<Object>> rows = new ArrayList<>();
        List<String> columns = new ArrayList<>();

        while (true) {
            if (page.containsKey("error")) {
                Map<?, ?> error = (Map<?, ?>) page.get("error");
                Object message = error.get("message");
                throw new IOException(message == null ? String.valueOf(error) : String.valueOf(message));
            }
            Object data = page.get("data");
            if (data instanceof List<?> dataRows) {
                for (Object dataRow : dataRows) {
                    rows.add(new ArrayList<>((List<Object>) dataRow));
                }
            }
            if (columns.isEmpty() && page.get("columns") instanceof List<?> columnRows) {
                for (Object columnRow : columnRows) {
                    Map<?, ?> column = (Map<?, ?>) columnRow;
                    columns.add(String.valueOf(column.get("name")));
                }
            }
            Object nextUri = page.get("nextUri");
            if (nextUri == null || String.valueOf(nextUri).isBlank()) {
                return new App.QueryResult(rows, columns);
            }

            HttpRequest.Builder builder = HttpRequest.newBuilder(URI.create(String.valueOf(nextUri)))
                    .timeout(java.time.Duration.ofSeconds(60))
                    .GET();
            applyHeaders(builder);
            HttpResponse<String> response = client.send(
                    builder.build(),
                    HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8)
            );
            if (response.statusCode() < 200 || response.statusCode() >= 300) {
                throw new IOException("trino next page failed: HTTP " + response.statusCode() + ": " + response.body());
            }
            page = App.JSON.readValue(response.body(), new TypeReference<>() {
            });
        }
    }

    private void applyHeaders(HttpRequest.Builder builder) {
        builder.header("X-Trino-User", config.trinoUser())
                .header("X-Trino-Source", "java-trino-log-search")
                .header("Content-Type", "text/plain; charset=utf-8");
        if (!config.trinoCatalog().isBlank()) {
            builder.header("X-Trino-Catalog", config.trinoCatalog());
        }
        if (!config.trinoSchema().isBlank()) {
            builder.header("X-Trino-Schema", config.trinoSchema());
        }
        if (!config.trinoPassword().isBlank()) {
            String token = Base64.getEncoder().encodeToString(
                    (config.trinoUser() + ":" + config.trinoPassword()).getBytes(StandardCharsets.UTF_8)
            );
            builder.header("Authorization", "Basic " + token);
        }
    }
}
